File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ExploreFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.BibleEntityType

/** The discovery surface (§7). Twin of iOS `ExploreFeature`. */
object ExploreFeature {
    data object State

    enum class Entry(val entityType: BibleEntityType?) {
        PEOPLE(BibleEntityType.PERSON), PLACES(BibleEntityType.PLACE), THEMES(BibleEntityType.THEME), EVENTS(BibleEntityType.EVENT), TIMELINE(null), BOOKS(null)
    }

    sealed interface Action {
        data class EntryTapped(val entry: Entry) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class Open(val entry: Entry) : DelegateAction
    }

    val reducer: Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            is Action.EntryTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.Open(action.entry))))
            is Action.Delegate -> state.only()
        }
    }
}

File: VerbumKit/Sources/Features/Explore/ExploreFeature.swift
import ComposableArchitecture
import Models

/// The discovery surface (spec §7): the paths into the knowledge graph, and
/// the timeline. The graph itself is entered from an entity (§21.4).
@Reducer
public struct ExploreFeature {
    @ObservableState
    public struct State: Equatable {
        public init() {}
    }

    public enum Entry: Equatable, Sendable, CaseIterable {
        case people, places, themes, events, timeline, books

        public var entityType: BibleEntityType? {
            switch self {
            case .people: .person
            case .places: .place
            case .themes: .theme
            case .events: .event
            case .timeline, .books: nil
            }
        }
    }

    public enum Action: Equatable {
        case entryTapped(Entry)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case open(Entry)
        }
    }

    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { _, action in
            switch action {
            case .entryTapped(let entry):
                return .send(.delegate(.open(entry)))
            case .delegate:
                return .none
            }
        }
    }
}
